service/common/src/main/java/org/apache/polaris/service/catalog/iceberg/IcebergCatalogHandler.java

```diff
@@ -762,7 +777,190 @@ public LoadTableResponse updateTable(
+    ops.commit((TableMetadata) null, builder.build());
+    return ops.current();
+  }
+
+  @VisibleForTesting
+  // TODO: Clean this up when CatalogHandler become extensible.
+  // Copy of CatalogHandler#commit
+  public static TableMetadata commit(TableOperations ops, UpdateTableRequest request) {
+    AtomicBoolean isRetry = new AtomicBoolean(false);
+
+    try {
+      Tasks.foreach(new TableOperations[] {ops})
+          .retry(4)
+          .exponentialBackoff(100L, 60000L, 1800000L, (double) 2.0F)
+          .onlyRetryOn(CommitFailedException.class)
+          .run(
+              (taskOps) -> {
+                TableMetadata base = isRetry.get() ? taskOps.refresh() : taskOps.current();
+                isRetry.set(true);
+                // Prev PR: https://github.com/apache/iceberg/pull/5888
+                boolean rollbackCompaction =
+                    PropertyUtil.propertyAsBoolean(
+                        taskOps.current().properties(), ROLLBACK_REPLACE_ENABLED_PROPERTY, false);
+
+                TableMetadata.Builder metadataBuilder = TableMetadata.buildFrom(base);
+                TableMetadata newBase = base;
+                try {
+                  request.requirements().forEach((requirement) -> requirement.validate(base));
+                } catch (CommitFailedException e) {
+                  if (!rollbackCompaction) {
+                    throw new ValidationFailureException(e);
+                  }
+                  // Since snapshot has already been created at the client end.
+                  // Nothing much can be done, we can move this
+                  // to writer specific thing, but it would be cool if catalog does this for us.
+                  // Inspect that the requirements states that snapshot
+                  // ref needs to be asserted this usually means in the update section
+                  // it has addSnapshot and setSnapshotRef
+                  UpdateRequirement.AssertRefSnapshotID addSnapshot = null;
+                  int found = 0;
+                  for (UpdateRequirement requirement : request.requirements()) {
+                    // there should be only add snapshot request
+                    if (requirement instanceof UpdateRequirement.AssertRefSnapshotID) {
+                      ++found;
+                      addSnapshot = (UpdateRequirement.AssertRefSnapshotID) requirement;
+                    }
+                  }
+
+                  if (found != 1) {
+                    // TODO: handle this case, find min snapshot id, to rollback to give it creates
+                    // lineage
+                    // lets not complicate things rn
+                    throw new ValidationFailureException(e);
+                  }
+
+                  Long parentSnapshotId = addSnapshot.snapshotId();
+                  // so we will first check all the snapshots on the top of
+                  // base on which the snapshot we want to commit is of type REPLACE ops.
```

Comment:
I think this block also probably should also get extracted into a private method. `findNoopSnapshots" or something